### DISCUSSION TITLE
New contributor interested in helping with kernel 4.19 → 6.1 upgrade (Rust support)

### DISCUSSION THREAD

**Discussion by Muresan73** [ev3dev/ev3dev#1653] [Q&A]
Hi ev3dev people! 👋

I've been playing around with ev3dev and noticed we're running on kernel 4.19. I'm interested in helping upgrade to 6.1, mainly because it would enable Rust kernel module bindings (which could open up some cool possibilities for the project i think).

**About me & my situation:**
- New to kernel development but eager to learn.
- My employer lets me spend 50% of my time on open source at the moment.
- Picked ev3dev because I love what you're doing here + would like to dip my toe with lower level linux staff.

**What I need help with:**
- Understanding what regression testing we'd need to ensure basic functionality
- Guidance on where to start with this upgrade process

I realize this is ambitious for a first-time kernel contributor. If there are better ways I could help out first or if this should be broken into smaller pieces, I'm open to suggestions. Just looking to contribute meaningfully while learning along the way.

Looking forward to your guidance!

**Comment by dlech** [comment]
A while ago, I started updating things for the 6.9 kernel. I didn't get very far though. There were some major changes in the rpmsg/remoteproc subsystems, so patches related to the PRU will need to be redone. Not exactly an easy task for someone new to the kernel, but I'm happy to answer any questions about it.

New branches missing or disabled  rproc stuff:
* https://github.com/ev3dev/ev3-kernel/commits/ev3dev-bookworm/
* https://github.com/ev3dev/lego-linux-drivers/tree/ev3dev-bookworm

Any patch from this branch that starts with `remoteproc:` or `rpmsg:` on top of v4.19 needs to be ported to the branch above.

**Reply by Muresan73** [reply]
While working on creating a Bookworm-based ev3dev image, I noticed there's currently no Debian Bookworm repository available for ev3dev packages.
Would it be possible to add Bookworm repository support? This would help streamline the process of building Bookworm-based images instead of manually building and installing packages.
I was working on getting the groundwork ready (working updated cross-compilation toolchain, kernel builds, and containers), but maintaining manual builds without a central repository is becoming cumbersome, so updating the CI could help ( plus having the community to review my approach and ensure I'm not going completely off track! ).

**Reply by dlech** [reply]
> Would it be possible to add Bookworm repository support?

There should be something there now.

**Comment by Muresan73** [comment]
Sorry if this is an uneducated question, but what is the logic or practice behind selecting specific commits to be copied to the ev3-kernel repository from the mainline Linux branch? I tried to separate the EV3-specific changes, and I was pleasantly surprised to see that the Bookworm branch already includes them. However, if I were to do this myself or wanted to upgrade it to a TLS version from an RC, I wouldn’t know how to approach it.

**Reply by dlech** [reply]
The logic is that we only pick up extra stuff that specifically fixes an issue for ev3dev users.

We always rebase our extra commits on top of a stable kernel tag so it should be easy to see what we have added on top of that. (`git rebase --onto <new tag> <old tag> ev3dev-bookworm`)

**Reply by Muresan73** [reply]
oh splendid, makes a lot of sense thank you :)

**Comment by Growflavor** [comment]
Posting this link & excerpt to help make it easier for I & others to read up on any progress, since I'm excited about the various current efforts on the 6.9 kernel :-)

https://github.com/ev3dev/ev3-kernel/pull/14
excerpt from above link: "With this fix and a few changes to get the lego-drivers to compile, can boot into brickman, although haven't (yet) tested whether motors and sensors work."
